symengine/dict.cpp

```diff
@@ -196,45 +196,78 @@ int map_int_Expr_compare(const map_int_Expr &A, const map_int_Expr &B)
     return 0;
 }
 
-template <class T>
-bool set_eq(const T &A, const T &B)
+bool multiset_basic_eq(const multiset_basic &a, const multiset_basic &b)
 {
-    // Can't be equal if # of entries differ:
-    if (A.size() != B.size())
-        return false;
-    // Loop over elements in "a" and "b":
-    auto a = A.begin();
-    auto b = B.begin();
-    for (; a != A.end(); ++a, ++b) {
-        if (neq(**a, **b))
-            return false; // values not equal
-    }
-    return true;
+    return set_eq<multiset_basic>(a, b);
 }
 
-template <class T>
-int set_compare(const T &A, const T &B)
+int multiset_basic_compare(const multiset_basic &a, const multiset_basic &b)
 {
-    if (A.size() != B.size())
-        return (A.size() < B.size()) ? -1 : 1;
-    auto a = A.begin();
-    auto b = B.begin();
-    int cmp;
-    for (; a != A.end(); ++a, ++b) {
-        cmp = (*a)->__cmp__(**b);
-        if (cmp != 0)
-            return cmp;
-    }
-    return 0;
+    return set_compare<multiset_basic>(a, b);
 }
 
-bool multiset_basic_eq(const multiset_basic &a, const multiset_basic &b)
+long mpz_hash(const integer_class z)
 {
-    return set_eq<multiset_basic>(a, b);
+    return mp_get_si(z);
 }
 
-int multiset_basic_compare(const multiset_basic &a, const multiset_basic &b)
+int umap_uvec_mpz_compare(const umap_uvec_mpz &a, const umap_uvec_mpz &b)
 {
-    return set_compare<multiset_basic>(a, b);
+    std::vector<vec_uint> va
+        = order_umap<vec_uint, umap_uvec_mpz, vec_uint_compare>(a);
+    std::vector<vec_uint> vb
+        = order_umap<vec_uint, umap_uvec_mpz, vec_uint_compare>(b);
+
+    if (va.empty())
+        if (!vb.empty())
```

Comment:
The block from line 221 to 229, can be removed, as it will be handled correctly without these checks too (will not enter the `for` loop)!